### PR TITLE
EI-1693 request recalculation of earnings for incentives

### PR DESCRIPTION
**Pull request by David-B-Read**

### CODE REVIEW

File: src/SFA.DAS.EmployerIncentives.Abstractions/DTOs/Commands/IncentiveLearnerIdentifierDto.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿namespace SFA.DAS.EmployerIncentives.Abstractions.DTOs.Commands
+{
+    public class IncentiveLearnerIdentifierDto
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Abstractions/DTOs/Commands/IncentiveLearnerIdentifierDto.cs:3]: We should raise a tech debt to remove the dto suffix and also move these classes out of the abstractions project.

File: src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs [hunk 2]
```diff
@@ -93,6 +94,36 @@ public void CalculateEarnings(CollectionCalendar collectionCalendar)
             Model.RefreshedLearnerForEarnings = false;
         }
 
+        public void RecalculateEarnings(CollectionCalendar collectionCalendar)
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs:97]: Do we need this?  Can't the existing CalculateEarnings be called there is duplication here.  Note collectionCalendar isn't used.

File: src/SFA.DAS.EmployerIncentives.Commands/RecalculateEarnings/RecalculateEarningsCommandHandler.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            foreach(var learnerIdentifier in command.IncentiveLearnerIdentifiers)
+            {
+                var incentive = await _domainRepository.FindByUlnWithinAccountLegalEntity(learnerIdentifier.ULN, learnerIdentifier.AccountLegalEntityId);
+
+                if (incentive == null)
+                {
+                    throw new ArgumentException($"Apprenticeship incentive not found for account legal entity {learnerIdentifier.AccountLegalEntityId} and ULN {learnerIdentifier.ULN}");
+                }
+
+                incentive.RecalculateEarnings(collectionCalendar);
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Commands/RecalculateEarnings/RecalculateEarningsCommandHandler.cs:35]: Just call CalculateEarnings ?

File: src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs [hunk 4]
```diff
@@ -93,6 +94,36 @@ public void CalculateEarnings(CollectionCalendar collectionCalendar)
...
+                return;
+            }
+
+            var recalculationRequired = false;
+            foreach(var pendingPayment in PendingPayments)
+            {
+                if (!ExistingPendingPaymentHasBeenPaid(pendingPayment))
+                {
+                    var existingPendingPaymentModel = pendingPayment.GetModel();
+                    Model.PendingPaymentModels.Remove(existingPendingPaymentModel);
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs:110]: No pending payment deleted event is thrown.

File: src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs [hunk 5]
```diff
@@ -93,6 +94,36 @@ public void CalculateEarnings(CollectionCalendar collectionCalendar)
...
+                    recalculationRequired = true;
+                }
+            }
+
+            if (recalculationRequired)
+            {
+                AddEvent(new EarningsRecalculationRequired
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs:117]: Why raise an event to do the calculation rather then call the method here?  (Although this method probably isn't required)

**Review by Stephek — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs (hunk 2 above)
Comment (David-B-Read) [reply src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs:97]: The requirement is to only recalculate earnings if a payment hasn't already been generated

File: src/SFA.DAS.EmployerIncentives.Abstractions/DTOs/Commands/IncentiveLearnerIdentifierDto.cs (hunk 1 above)
Comment (David-B-Read) [reply src/SFA.DAS.EmployerIncentives.Abstractions/DTOs/Commands/IncentiveLearnerIdentifierDto.cs:3]: Created ticket EI-1715

File: src/SFA.DAS.EmployerIncentives.Api/Controllers/ApprenticeshipIncentiveCommandController.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace SFA.DAS.EmployerIncentives.Api.Controllers
+{
+    public class ApprenticeshipIncentiveCommandController : ApiCommandControllerBase
+    {
+        public ApprenticeshipIncentiveCommandController(ICommandDispatcher commandDispatcher) : base(commandDispatcher)
+        {
+        }
+
+        [HttpPost("/recalculateEarnings")]
```
[the remark this answers is not in the document]
Comment (David-B-Read) [reply src/SFA.DAS.EmployerIncentives.Api/Controllers/ApprenticeshipIncentiveCommandController.cs]: Updated

File: src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs (hunk 4 above)
Comment (David-B-Read) [reply src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs]: That's been added

File: src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs (hunk 5 above)
Comment (David-B-Read) [reply src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs]: Events have been removed, calling CalculateEarnings directly

**Review by Stephek — APPROVED** [review]

**Pull request merged**